network/stub/network.go

```diff
@@ -268,23 +268,35 @@ func (n *Network) sendToAllTargets(m *PendingMessage, syncOnProcess bool) error
 		receiverNetwork.seen(key)
 
 		// finds the engine of the targeted Network
-		receiverEngine, ok := receiverNetwork.engines[m.Channel]
-		if !ok {
-			return fmt.Errorf("could find engine ID: %v for node: %v", m.Channel, nodeID)
+		err := receiverNetwork.processWithEngine(receiverNetwork == n, syncOnProcess, m)
```

Comment:
there are multiple issues here.

1. haveSeen & seen should be part of processWithEngine (this currently is a threadsafe data race)
2. this will deadlock.  you need to acquire the receiverNetworks locks in some consistent order or independently to avoid the deadlocking.

I would make probably Hub threadsafe with it's own internal lock, remove top level lock in this function, and consistently grab lock inside processWithEngine instead (also switching to RWMutex).

Also split this out into a different PR since it's not simple

Reply:
Ok. Factored jobqueue into https://github.com/onflow/flow-go/pull/3092. Will work on reworking the mocked network.